apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/Interval.cs

```diff
@@ -0,0 +1,381 @@
+    /// </summary>
+    public int Days { get; }
+
+    /// <summary>
+    /// Nanosecond portion of this interval.
+    /// </summary>
+    public BigInteger Nanoseconds { get; }
+
+    private Interval(int months, int days, BigInteger nanoseconds)
+    {
+        Months = GaxPreconditions.CheckArgumentRange(months, nameof(months), MinMonths, MaxMonths);
+        Days = GaxPreconditions.CheckArgumentRange(days, nameof(days), MinDays, MaxDays);
+        Nanoseconds = GaxPreconditions.CheckArgumentRange(nanoseconds, nameof(nanoseconds), MinNanoseconds, MaxNanoseconds);
+    }
+
+    /// <summary>
+    /// Creates an interval from the given months, days and nanoseconds.
+    /// </summary>
+    public static Interval FromMonthsDaysNanos(int months, int days, BigInteger nanos) => new Interval(months, days, nanos);
+
+    /// <summary>
+    /// Creates an interval from the given months.
+    /// </summary>
+    public static Interval FromMonths(int months) => new Interval(months, 0, 0);
+
+    /// <summary>
+    /// Creates an interval from the given days.
+    /// </summary>
+    public static Interval FromDays(int days) => new Interval(0, days, 0);
+
+    /// <summary>
+    /// Creates an interval from the given seconds.
+    /// </summary>
+    public static Interval FromSeconds(decimal seconds) => new Interval(0, 0, SecondsToNanoseconds(seconds));
+
+    /// <summary>
+    /// Creates an interval from the given milliseconds.
+    /// </summary>
+    public static Interval FromMilliseconds(BigInteger milliseconds) => new Interval(0, 0, milliseconds * NanosecondsInAMillisecond);
+
+    /// <summary>
+    /// Creates an interval from the given microseconds.
+    /// </summary>
+    public static Interval FromMicroseconds(BigInteger microseconds) => new Interval(0, 0, microseconds * NanosecondsInAMicrosecond);
+
+    /// <summary>
+    /// Creates an interval from the given nanoseconds.
+    /// </summary>
+    public static Interval FromNanoseconds(BigInteger nanoseconds) => new Interval(0, 0, nanoseconds);
+
+    /// <summary>
+    /// Parses the string represantion in ISO8601 format of a duration returning its Interval equivalent.
+    /// </summary>
+    /// <param name="text">
+    /// The string representation of a duration in ISO8601 format: P[n]Y[n]M[n]DT[n]H[n]M[n[.fraction]]S.
+    /// </param>
+    public static Interval Parse(string text)
+    {
+        GaxPreconditions.CheckNotNull(text, nameof(text));
+
+        var state = new IntervalParsingState();
+        int end;
+        do
+        {
+            end = text.IndexOfAny(state.NextAllowed, state.Start);
+
+            // We couldn't find any of the allowed characters of which we needed to find one.
+            if (end == -1)
+            {
+                throw new FormatException($"Expected one of {string.Join(", ", state.NextAllowed)} in {text.Substring(state.Start)}.");
+            }
+            // P[n]Y[n]M[n]DT[n]H[n]M[n[.fraction]]S
```

Comment:
Are we happy with *any* aspect of this being negative independent of other aspects, e.g. `P5H-2.2S`?

Additionally, the text `P[n]Y[n]M[n]DT[n]H[n]M[n[.fraction]]S` in the comment earlier doesn't make it clear what can be included/omitted. Maybe some examples would help.